### PR TITLE
Check for gcc and bump build number in compiler migration and track host requirements

### LINKED ISSUE #215: Check for gcc in Compiler upgrade
Some older recipes  and R recipes use gcc instead of toolchain. Let's ignore R recipes for now, since there are some issues with conda-build we have to figure out. We can migrate recipes that are not R, but use gcc using `conda smithy update-cb3`

### LINKED ISSUE #230: Build bump for compiler syntax
@justcalamari Would it be possible to put a build number bump in the compiler syntax migrator.

### PR DESCRIPTION
**Pull request by justcalamari**
Closes #215
Closes #230

### CODE REVIEW

File: conda_forge_tick/migrators.xsh [hunk 1]
```diff
@@ -335,9 +335,11 @@ class Compiler(Migrator):
 
     rerender = True
 
+    compilers = ['toolchain', 'gcc']
+
     def filter(self, attrs):
         conditional = super().filter(attrs)
-        return conditional or 'toolchain' not in attrs.get('req', [])
+        return conditional or not any(x in attrs.get('req', []) for x in self.compilers)
```
Comment (isuruf) [inline conda_forge_tick/migrators.xsh]: We should not do this for R recipes as there are some conda-build issue we should figure out

File: conda_forge_tick/migrators.xsh [hunk 2]
```diff
@@ -335,9 +335,11 @@ class Compiler(Migrator):
 
     rerender = True
 
+    compilers = ['toolchain', 'gcc']
```
Comment (scopatz) [inline conda_forge_tick/migrators.xsh]: This should probably be a set

File: conda_forge_tick/migrators.xsh [hunk 3]
```diff
@@ -336,6 +336,12 @@ class Compiler(Migrator):
     rerender = True
 
     compilers = {'toolchain', 'gcc'}
+    patterns = (('meta.yaml', '  number:\s*[0-9]+', '  number: $BUILDNUM'),
```
Comment (CJ-Wright) [inline conda_forge_tick/migrators.xsh]: Thank you!

Comment (CJ-Wright) [reply conda_forge_tick/migrators.xsh]: Would it be possible to put this everywhere else? (Not version bump but I think all the others need this too (noarch, JS).

File: conda_forge_tick/migrators.xsh [hunk 4]
```diff
@@ -346,6 +352,12 @@ class Compiler(Migrator):
 
     def migrate(self, recipe_dir, attrs, **kwargs):
         self.out = $(conda-smithy update-cb3 --recipe_directory @(recipe_dir))
+        buildnum = attrs["build"]["number"] + 1
```
Comment (CJ-Wright) [inline conda_forge_tick/migrators.xsh]: This might not be ok, since this number will be 200+ for blas builds.

Comment (CJ-Wright) [reply conda_forge_tick/migrators.xsh]: Might need to pull this from `build_number` jinja2 syntax or from attrs.

**Comment by CJ-Wright** [comment]
Thank you very much @justcalamari

**Pull request merged**